File: backend/src/routes/exportData.ts
import { Router } from 'express';
import type { BackupPayload, Op, ExitPrices } from '../types';

export const exportRouter = Router();

// GET /api/export — generates the authenticated user's full backup JSON
exportRouter.get('/', async (req, res) => {
  const [opsResult, exitPricesResult] = await Promise.all([
    req.supabase!.from('ops').select('*').order('date', { ascending: true }),
    req.supabase!.from('exit_prices').select('coin_id, exit_price'),
  ]);

  if (opsResult.error) {
    res.status(500).json({ error: opsResult.error.message });
    return;
  }
  if (exitPricesResult.error) {
    res.status(500).json({ error: exitPricesResult.error.message });
    return;
  }

  const ops: Op[] = opsResult.data.map((row) => ({
    id: row.id,
    date: row.date,
    coinId: row.coin_id,
    symbol: row.symbol,
    name: row.name,
    type: row.type,
    qty: Number(row.qty),
    price: Number(row.price),
    fee: Number(row.fee),
    total: Number(row.total),
    platform: row.platform,
  }));

  const exitPrices: ExitPrices = {};
  for (const row of exitPricesResult.data) exitPrices[row.coin_id] = Number(row.exit_price);

  const payload: BackupPayload = {
    version: 1,
    exportedAt: new Date().toISOString(),
    ops,
    exitPrices,
  };

  res.json(payload);
});

File: backend/app/routes/export_data.py
import datetime
from fastapi import APIRouter, Depends, HTTPException, status
from app.dependencies import AuthContext, require_auth
from app.db.postgres_client import get_conn
from app.models import BackupPayload, Op

router = APIRouter()

_SELECT = "id, date, coin_id, symbol, name, type, qty, price, fee, total, platform"


@router.get("", response_model=BackupPayload)
def export_backup(auth: AuthContext = Depends(require_auth)):
    conn = get_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                f"SELECT {_SELECT} FROM ops WHERE user_id = %s ORDER BY date",
                (auth.user_id,),
            )
            ops_rows = cur.fetchall()

            cur.execute(
                "SELECT coin_id, exit_price FROM exit_prices WHERE user_id = %s",
                (auth.user_id,),
            )
            exit_rows = cur.fetchall()
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

    ops = [
        Op(
            id=str(r["id"]),
            date=str(r["date"]),
            coinId=r["coin_id"],
            symbol=r["symbol"],
            name=r["name"],
            type=r["type"],
            qty=float(r["qty"]),
            price=float(r["price"]),
            fee=float(r["fee"]),
            total=float(r["total"]),
            platform=r["platform"],
        )
        for r in ops_rows
    ]
    exit_prices = {r["coin_id"]: float(r["exit_price"]) for r in exit_rows}

    return BackupPayload(
        version=1,
        exportedAt=datetime.datetime.now(datetime.timezone.utc).isoformat(),
        ops=ops,
        exitPrices=exit_prices,
    )
